opentelemetry-sdk/src/opentelemetry/sdk/logs/__init__.py

```diff
@@ -244,20 +244,63 @@ def force_flush(self, timeout_millis: int = 30000) -> bool:
         return True
 
 
+# skip natural LogRecord attributes
+# http://docs.python.org/library/logging.html#logrecord-attributes
+_RESERVED_ATTRS = frozenset(
+    (
+        "args",
+        "asctime",
+        "created",
+        "exc_info",
+        "exc_text",
+        "filename",
+        "funcName",
+        "levelname",
+        "levelno",
+        "lineno",
+        "module",
+        "msecs",
+        "message",
+        "msg",
+        "name",
+        "pathname",
+        "process",
+        "processName",
```

Comment:
Many of these attributes enrich the data and are important enough to be not discarded.

Reply:
They aren't being discarded, just not included by default. https://github.com/open-telemetry/opentelemetry-python/pull/1952#discussion_r669284823 would be a way to let users include them.

Reply:
Should have been more clear, why is this the default choice?. We may eventually go with this but I want to point out that this is a prototype and we want to release an experimental package and then gather feedback from more audiences.

Reply:
Ah that makes sense.

So for background, the way I see it there are two issues at play:
1. Any user injected data (via the `extra` keyword argument) gets merged into `LogRecord.__dict__`, so there is no way to tell apart user defined data from `LogRecord`'s own data. This list of attributes corresponds to `LogRecord`'s documented attributes and allows pulling out the user defined attributes.
2. Key collisions. As discussed in the other comment, you can't set a user defined attribute that collides with one of the above, e.g. `process`. There are several ways to deal with this, including a prefix (which can then be stripped), a nested key to be extracted/flattened or any other similar approach.

In the most basic form of this implementation, I think it makes sense to let users pass in their custom data and have them attached to the OTLP `LogRecord` as OTLP Attributes. To meet that requirement, we need to be able to filter out `LogRecord`'s own attributes.

That said, I do hear your point on trying to keep it simple.
How about I remove the `attributes_key` functionality (users can just use a prefix, as you point out) and we go with this minimal implementation.

Later, based on user feedback, options can be added to do fancier stuff like include `LogRecord`'s own attributes (maybe even include some by default), rename keys, auto remove prefixes, etc. All of this should be doable in a backwards compatible manner.

Does this sound reasonable?

Reply:
Just for the record, those are the documented attributes, but not necessarily all the attributes of `LogRecord` that may suffer a collision. A way to find them all is to check directly for the actual attributes in a `LogRecord` instance.

Here is a complicated (:stuck_out_tongue_closed_eyes:) way of getting them without knowing beforehand how many parameters `LogRecord.__init__` requires:

```python
from inspect import signature, Parameter
from logging import LogRecord


[
    "asctime",
    *[
        attribute for attribute in dir(
            LogRecord(
                *[None] * (
                    len(
                        [
                            parameter
                            for parameter in
                            signature(
                                LogRecord.__init__
                            ).parameters.values() if
                            parameter.kind == Parameter.POSITIONAL_ONLY or
                            parameter.kind == Parameter.POSITIONAL_OR_KEYWORD
                        ]
                    ) - 1  # because self is also a positional attribute
                )
            )
        ) if not attribute.startswith("_")
    ]
]
```
Notice how an instance does not have an `asctime` attribute (it gets one when it has been formatted by a `logging.Formatter` object).

Reply:
Wow ya that's a bit complex! I feel like there are maybe simpler versions? Eg https://stackoverflow.com/a/59934509/6582418

I was just going with what `python-json-logging` uses, which given that it is widely used I figured would suffice.

Reply:
The script above and the one you linked are actually very similar. Both instantiate a `LogRecord` object and then use it to find which attributes it has. The advantage of the script above is that it does not require to know beforehand what to pass to `__init__` so it saves you from the very unlikely scenario that its interface changes and it also saves you from having to type `'',0,None,0,'',[],None`.

:smile:

Reply:
Makes sense. You're the expert here so it's up to you where to set the complexity/interpretability vs. functionality tradeoff 😄 . If you want me to go with the above instead of the hardcoded values, just say so and I'll put it in 👍

Reply:
Well, to be honest, I don't think is worth the complication, hardcoded values are fine. I just wanted to point out that there are other attributes that can cause collisions. Here is the result of running my script, if you want to take them into consideration. :+1:

`['asctime', 'args', 'created', 'exc_info', 'exc_text', 'filename', 'funcName', 'getMessage', 'levelname', 'levelno', 'lineno', 'module', 'msecs', 'msg', 'name', 'pathname', 'process', 'processName', 'relativeCreated', 'stack_info', 'thread', 'threadName']`

Reply:
Pushed your list @ocelotl . Thanks!

Reply:
The reason I am still not in favor of this hardcoded `_RESERVED_ATTRS` is I have seen various ways people configure logging to include many of these attrs. It can be as simple as default `"%(levelname)s:%(name)s:%(message)s"`  or customised `"[%(asctime)s] %(levelname)s [%(name)s.%(funcName)s:%(lineno)d]%(process:processName)s"`. This data helps in narrowing down really quickly and I believe we shouldn't be discarding these standard attributes from the log records by default. I suggest we include them in OTLP log record attributes.

Sidenote: What's with `getMessage`? It is a method on LogRecord class.

Reply:
> I have seen various ways people configure logging to include many of these attrs

Are you referring to configuring the standard library logging? If so, that would be done by a formatter, which runs before this Handler. So they would just be part of the log message at this point.

> I suggest we include them in OTLP log record attributes

I don't think you want to or even can include them all by default. `exc_info` for example is a random Python object (a traceback instance I believe), so I suspect it won't serialize nicely. Unless I'm missing something about how wire serialization works?

Another option is to use the composability of the logging module. Users can add a _Filter_ on this handler that gives them the ability to rename attributes.

Reply:
> I don't think you want to or even can include them all by default. exc_info for example is a random Python object (a traceback instance I believe), so I suspect it won't serialize nicely. Unless I'm missing something about how wire serialization works?

we may not be able to add everything but many of these attributes are trivial. We do some traceback related serialization in tracing so I assume similar things can be done here https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L875

> Users can add a Filter on this handler that gives them the ability to rename attributes.

I am not sure I understand what do you mean by `ability to rename attributes`? Can you elaborate on this?

Reply:
I'm just trying to find a way to solve the problem using the logging module's built in composability. I pushed an example in 5765374

Reply:
That's a different problem right? As I pointed out earlier you could avoid the attributes collision by namespacing naming we usually do in semantic conventions. I want to reiterate that let's keep things simple and make any kind of decisions later in the process once we have enough feedback from the users.

Reply:
> we shouldn't be discarding these standard attributes from the log records by default

I was trying to give you a solution to the problem you brought up, but I guess I'm not understanding the problem. Would you mind giving me a codified example of the behavior you would like that this PR breaks/blocks?

>  let's keep things simple and make any kind of decisions later in the process once we have enough feedback from the users

Could you help me understand what this means? I am making this PR as a potential user. Without this functionally, I would not be able to use this feature.

Reply:
I appreciate you taking the time to put this out and understand the importance of having attributes populated, what I am trying to ask is what is the reason to not add standard attributes from the log records by default here `k: v for k, v in vars(record).items() if k not in _RESERVED_ATTRS`? logging record attributes such as filename, processName, etc.. can be OTLP log record attributes along with user defined attributes.

Reply:
Several reasons to not include them by default:
1. Not all of them can easily be serialized.
2. Not all users may want all of them. So you're left adding a feature to exclude them. Which is the same as a feature to include them.
3. Adding some (or all) of the attributes down the road can be done without breaking existing code in the wild. The converse is not possible.

Besides, if user _really_ wanted them in there, they can do so _without any changes/features_ from OpenTelemetry. They can use this PR as is, as demonstrated in 5765374. The only thing they can't do is add these attributes under the same name they have in the `LogRecord` class. But I doubt the semantic convention is `funcName`, so that isn't really a problem.

So overall, if the goal is to punt decisions down the road (as it always should be), I think not adding them is miles safer than the alternative.

Reply:
On second thought I think the discussion drifted away since you opened the PR. I am fine with this PR not bothering about standard attributes from python log records.